lib/factor-rules.js

```diff
@@ -0,0 +1,373 @@
+  }
+  if (num < 0 && (Math.abs(root) % 2 !== 1)) {
+    throw new Error('Root must be odd when a is negative.')
+  }
+
+  // Edge cases zero and infinity.
+  // e.g 0^3 = 0, 0^-3 = Infinity
+  if (num === 0) {
+    return inv ? Infinity : 0
+  }
+
+  if (num === 1) {
+    return 1
+  }
+
+  if (!isFinite(num)) {
+    return inv ? 0 : num
+  }
+
+  // Source: https://rosettacode.org/wiki/Nth_root#JavaScript
+  const n = root
+  const prec = precision
+
+  let x = 1 // Initial guess
+  for (let i = 0 ; i < prec ; i++) {
+    x = 1 / n * ((n - 1) * x + (num / Math.pow(x, n - 1)))
+  }
+
+  return inv ? 1 / x : x
+}
+
+// TODO: handle fractional coefficients
+// TODO: isPolynomialTerm should handle fractional coeffs
+
+// e.g. 4x^2 + 12x + 9 -> (2x + 3)^2
+export const FACTOR_PERFECT_SQUARE_TRINOMIALS = defineRule(
+  (node) => {
+    const isFactorable = canApplyRule(FACTOR_SUM_PRODUCT_RULE, node)
+    if (isFactorable) {
+      const result = applyRule(FACTOR_SUM_PRODUCT_RULE, node)
+      return print(result.args[0]) === print(result.args[1])
+        ? {node} : null
+    }
+  },
+
+  (node) => {
+    const result = applyRule(FACTOR_SUM_PRODUCT_RULE, node)
+    return build.pow(result.args[0], build.number(2))
+  }
+)
+
+const getExponent = (node) => {
+    if (query.isNumber(node)) {
+        return 0
+    } else if (query.isIdentifier(node)){
+        return 1
+    } else if (query.isPow(node)) {
+        return query.getValue(node.args[1])
+    } else if (query.isMul(node)){
+        return getExponent(node.args[1])
+    } else if (query.isNeg(node)) {
+        const variable = node.args[0]
+        return getExponent(variable.args[1])
+    } else {
+        return null
+    }
+}
+
+// e.g. 2 + 3x^2 + 3x - 4x^3 -> -4x^3 + 3x^2 + 3x + 2
+export const REARRANGE_TERMS = defineRule(
+    (node) => {
+        if (query.isAdd(node)) {
+            return node.args.some(arg => {
+                return isPolynomialTerm(arg)
+            }) ? {node} : null
+        }
+    },
+
+    (node) => {
+        const ordered = node.args.sort(function(a,b) {return getExponent(b) - getExponent(a)})
+        return build.add(...ordered)
+    }
+)
+
+const getFactors = (num) => {
+  var factors = []
+
+  for(var i = 1; i <= Math.sqrt(num); i++){
+    if(Number.isInteger(num/i)){
+      factors.push([i, num/i])
+    }
+  }
+
+  return factors
+}
+
+/*
+  Cross factoring method 
+
+  Example: (6x^2 + 7x + 2)
+
+  1. Get factors for 6 and 2
+  6: [1,6], [2,3]    2: [1,2]
+
+  2. For each pair of factors (one from the first
+  and one from the last term) ...
+  [1,6] and [1,2]      |       [2,3] and [1,2]
+
+  3. Cross multiply the factors and check if
+  the result matches one of the eight conditions
+  (the second factor shld also multiply to equal the thirdCoef)
+  [1,6] and [1,2]
+
+  Check: 1 * 2 + 6 * 1 ?= 7 so on and so forth
+  Once we reach 2 * 2 + 3 * 1 ?= 7, we have found a match.
+  (Note: 3 * 1 = 3 satisfying the second condition)
```

Comment:
I'm confused about this note.... don't we want to check if the multiple the first candidate coefficients results in the first coefficient of the trinomial and multiplying the both of the second candidate numbers results in the constant of the trinomial.  In this case 2 * 3 = 6 and 2 * 1 = 2 so it works out.